checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,845 @@
- * inference logic for owning annotations on final owning fields. It adds an @Owning annotation on a
- * field if it finds a method that satisfies the @MustCall obligation of the field along some path
- * to the regular exit point.
+ * This class implements the annotation inference algorithm for the Resource Leak Checker. It infers
+ * annotations such as {@code @}{@link Owning} on owning fields and
+ * parameters, @EnsuresCalledMethods on methods, and @InheritableMustCall on class declarations.
+ *
+ * <p>Each instance of this class corresponds to a single control flow graph (CFG), typically
+ * representing a method.
+ *
+ * <p>The algorithm determines if the @MustCall obligation of a field is satisfied along some path
+ * leading to the regular exit point of the method. If the obligation is satisfied, it adds
+ * an @Owning annotation on the field and an @EnsuresCalledMethods annotation on the method being
+ * analyzed. Additionally, if the method being analyzed satisfies the must-call obligation of all
+ * the enclosed owning fields, it adds a @InheritableMustCall annotation on the enclosing class.
+ *
+ * <p>Note: This class makes the assumption that the must-call set has only one element. Must-call
+ * sets with more than one element may be supported in the future.
+ *
+ * <p>See {@link ResourceLeakChecker#ENABLE_WPI_FOR_RLC} for an explanation of the meaning of the
+ * flags {@code -Ainfer} and {@code -AenableWpiForRlc}.
+ *
+ * @see <a
+ *     href="https://checkerframework.org/manual/#resource-leak-checker-inference-algo">Automatic
+ *     Inference of Resource Leak Specifications</a>
  */
 public class MustCallInference {
 
-  /** The set of owning fields. */
+  /**
+   * The fields that have been inferred to be disposed within the CFG currently under analysis. All
+   * of these fields will be given an @Owning annotation.
+   */
+  private final Set<VariableElement> disposedFields = new HashSet<>();
+
+  /**
+   * The fields with written or inferred {@code @Owning} annotations at the entry point of the CFG
+   * currently under analysis.
+   */
   private final Set<VariableElement> owningFields = new HashSet<>();
 
   /**
    * The type factory for the Resource Leak Checker, which is used to access the Must Call Checker.
    */
-  private final ResourceLeakAnnotatedTypeFactory typeFactory;
+  private final ResourceLeakAnnotatedTypeFactory resourceLeakAtf;
+
+  /** The MustCallConsistencyAnalyzer. */
+  private final MustCallConsistencyAnalyzer mcca;
 
   /** The {@link Owning} annotation. */
   protected final AnnotationMirror OWNING;
 
-  /** The control flow graph. */
+  /**
+   * The control flow graph of the current method. There is a separate MustCallInference for each
+   * method.
+   */
   private final ControlFlowGraph cfg;
 
+  /** The MethodTree of the current method. */
+  private final MethodTree methodTree;
+
+  /** The element for the current method. */
+  private final ExecutableElement methodElt;
```

Comment:
@mernst does the Nullness Checker run on this code as part of CI?  If so, I'm surprised it does not report an issue for this line.  Line 140 assigns this field the result of `TreeUtils.elementFromDeclaration()` which is declared as returning `@Nullable`.

Reply:
The checkNullness task runs on:
  * All files outside the 'framework' and 'checker' modules.
  * In the 'framework' and 'checker' modules, files with `@AnnotatedFor("nullness")`.

So, the file `MustCallInference.java` is not type-checked for nullness by default, but the file can opt into such checking.